base/broadcast.jl

```diff
@@ -257,13 +257,23 @@ end
 @inline broadcast_elwise_op(f, As...) =
     broadcast!(f, similar(Array{promote_eltype_op(f, As...)}, broadcast_indices(As...)), As...)
 
-@pure typestuple(a) = Tuple{eltype(a)}
-@pure typestuple(T::Type) = Tuple{Type{T}}
-@pure typestuple(a, b...) = Tuple{typestuple(a).types..., typestuple(b...).types...}
+typestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
+typestuple(T::Type) = (Base.@_pure_meta; Tuple{Type{T}})
+typestuple(a, b...) = (Base.@_pure_meta; Tuple{typestuple(a).types..., typestuple(b...).types...})
+
+ftype(f, A) = typeof(a->f(a))
```

Comment:
@vtjnash, could you expand briefly? IIUC, you mean that `ftype(f, A) = Core.Typeof(f)` could replace `ftype(f, A) = typeof(a -> f(a))` whereas `ftype(f, A) = typeof(f)` could not,  as `Core.Typeof(f)` will return `Type{T}` when `f::T<:Type` whereas `typeof(f)` will not. Is that correct? If so, would defining `ftype(f, A) = typeof(f); ftype(f::Type, A) = Type{T}` as in the early version of #19421 (see https://github.com/JuliaLang/julia/pull/19633#discussion_r92925738) or instead `ftype(f, A) = Core.Typeof(f)` be better in this case? Thanks!